catalog/resource_grant.go

```diff
@@ -0,0 +1,187 @@
+			remote[v.Principal] = util.SliceToSet(v.Privileges)
+		}
+	}
+	// STEP 1: detect overlaps
+	for principal, confPrivs := range configured {
+		remotePrivs, ok := remote[principal]
+		if !ok {
+			remotePrivs = util.SliceToSet([]string{})
+		}
+		add := util.SetToSlice(confPrivs.Difference(remotePrivs))
+		remove := util.SetToSlice(remotePrivs.Difference(confPrivs))
+		if len(add) == 0 && len(remove) == 0 {
+			continue
+		}
+		diff.Changes = append(diff.Changes, permissions.UnityCatalogPermissionsChange{
+			Principal: principal,
+			Add:       add,
+			Remove:    remove,
+		})
+	}
+	// STEP 2: non overlap - simply remove
+	for principal, remove := range remote {
+		_, ok := configured[principal]
+		if ok { // already handled in STEP 1
+			continue
+		}
+		diff.Changes = append(diff.Changes, permissions.UnityCatalogPermissionsChange{
+			Principal: principal,
+			Remove:    util.SetToSlice(remove),
+		})
+	}
+	// so that we can deterministic tests
+	sort.Slice(diff.Changes, func(i, j int) bool {
+		return diff.Changes[i].Principal < diff.Changes[j].Principal
+	})
+	return diff
+}
+
+// replacePermissionsForPrincipal merges removal diff of existing permissions on the platform
+func replacePermissionsForPrincipal(a permissions.UnityCatalogPermissionsAPI, securable string, name string, principal string, list permissions.UnityCatalogPermissionsList) error {
+	existing, err := a.GetPermissions(securable, name)
+	if err != nil {
+		return err
+	}
+	return a.UpdatePermissions(securable, name, diffPermissionsForPrincipal(principal, list, existing))
+}
+
+func filterPermissionsForPrincipal(in permissions.UnityCatalogPermissionsList, principal string) (out permissions.UnityCatalogPermissionsList) {
+
+	for _, v := range in.Assignments {
+		if v.Principal == principal {
+			out.Assignments = append(out.Assignments, v)
+		}
+	}
+	return
+}
+
+func ResourceGrant() *schema.Resource {
+	s := map[string]*schema.Schema{
+		"principal": {
+			Type:     schema.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
+		"privileges": {
+			Type:     schema.TypeSet,
+			Required: true,
+			Elem:     &schema.Schema{Type: schema.TypeString},
+			Set:      schema.HashString,
+		},
+	}
+	// Handle all of the securable resource types
+	allFields := []string{}
+	for field := range permissions.Mappings {
+		allFields = append(allFields, field)
+	}
+
+	for _, field := range allFields {
+		s[field] = &schema.Schema{
+			Type:          schema.TypeString,
+			Optional:      true,
+			ForceNew:      true,
+			ConflictsWith: util.SliceWithoutString(allFields, field),
+		}
+	}
+
+	return common.Resource{
+		Schema: s,
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
+			if d.Id() == "" {
+				// unfortunately we cannot do validation before dependent resources exist with tfsdkv2
+				return nil
+			}
+			var grants permissions.UnityCatalogPermissionsList
+			common.DiffToStructPointer(d, s, &grants)
+			return permissions.Mappings.Validate(d, grants)
```

Comment:
we should let server side validation to kick in

Reply:
I matched the impl to that which exists in `resource_grants`. Should I remove it there also for consistency?

Reply:
we should leave it as is for `resource_grants` for backwards compatibility, but should avoid it for the new resource

Reply:
Done